src/librbd/mirror_snapshot/CreateRequest.cc

```diff
@@ -0,0 +1,390 @@
+    r = cls_client::mirror_image_get_finish(&iter, &mirror_image);
+  }
+
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to retrieve mirroring state: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  if (mirror_image.state != cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+    lderr(cct) << "mirroring is disabled" << dendl;
+    finish(-EINVAL);
+    return;
+  }
+
+  // XXX: should the caller be responsible for this validation?
+  if (!validate_snapshot_type()) {
+    finish(-EINVAL);
+    return;
+  }
+
+  uuid_d uuid_gen;
+  uuid_gen.generate_random();
+  m_snap_name = ".mirror." + mirror_image.global_image_id + "." +
+    uuid_gen.to_string();
+
+  get_mirror_uuid();
+}
+
+template <typename I>
+void CreateRequest<I>::get_mirror_uuid() {
+  if (!m_mirror_uuid.empty()) {
+    get_mirror_peers();
+  }
+
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_uuid_get_start(&op);
+
+  librados::AioCompletion *comp = create_rados_callback<
+    CreateRequest<I>, &CreateRequest<I>::handle_get_mirror_uuid>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_out_bl);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void CreateRequest<I>::handle_get_mirror_uuid(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r == 0) {
+    auto iter = m_out_bl.cbegin();
+    r = cls_client::mirror_uuid_get_finish(&iter, &m_mirror_uuid);
+  }
+
+  if (r < 0) {
+    lderr(cct) << "failed to retrieve local mirror uuid: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  get_mirror_peers();
+}
+
+template <typename I>
+void CreateRequest<I>::get_mirror_peers() {
+  if (m_type == cls::rbd::MIRROR_SNAPSHOT_TYPE_NON_PRIMARY) {
+    create_snapshot();
+    return;
+  }
+
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_peer_list_start(&op);
+
+  librados::AioCompletion *comp = create_rados_callback<
+    CreateRequest<I>, &CreateRequest<I>::handle_get_mirror_peers>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_out_bl);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void CreateRequest<I>::handle_get_mirror_peers(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  std::vector<cls::rbd::MirrorPeer> peers;
+  if (r == 0) {
+    auto iter = m_out_bl.cbegin();
+    r = cls_client::mirror_peer_list_finish(&iter, &peers);
+  }
+
+  if (r < 0) {
+    lderr(cct) << "failed to retrieve mirror peers: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  for (auto &peer : peers) {
+    // XXX: filter out by direction?
+    m_mirror_peers.insert(peer.uuid);
```

Comment:
How will the remote peer know its local peer mirror peer uuid? I had to use the site name as a match for the mirror status changes.

Reply:
The ImageReplayer would also use site name to match the peer uuid, i.e. get its site name with `rbd.mirror_site_name_get`, get remote peers with `cls_client::mirror_peer_list` and find that matches site name. Or do you think we can just store the peer site name here instead of uuid?

Reply:
That might be overly complicated to force it to list the remote peers to hit a match for each ImageReplayer. Using the site name would at least avoid that need w/ the caveat that changing your site-name could potentially cause a site to become unlinked from the primary image (could also happen if you delete/re-add the peer). We could also use the fsid as the match since we are also storing that in the peers (remote fsid is stored in the local peer upon the periodic ping).